apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -46,126 +48,259 @@ interface EventData {
+    return { success: false, phase, error: originalError, organizerId };
+  } else {
+    throw new Error(`originalError is not an error`);
+  }
+}
+
 /**
- * Responsible for syncing users from Pretix into an internal representation.
+ * We have a concept of a sync "run", which occurs on an interval.
+ * A run can fail, can complete synchronization, or can be suspended due
+ * to rate-limiting on requests to Pretix.
+ *
+ * In order to do a full synchronization for a given organizer, multiple
+ * runs may be required if rate-limiting makes it impossible to fetch
+ * all of the necessary data in a single run.
  */
-export class DevconnectPretixSyncService {
-  private static readonly SYNC_INTERVAL_MS = 1000 * 60;
+interface SyncRunSuccess {
+  success: true;
+  organizerId: string;
+  outcome: "complete" | "rate-limited";
+}
+
+type SyncResult = SyncRunSuccess | SyncErrorCause;
+
+/**
+ * This class implements a rate-limited fetcher. It provides a function
+ * which can be passed to the Pretix fetch functions, in place of the
+ * native 'fetch' API call. Behind the scenes, this function will
+ * increment a counter. When the counter exceeds the limit, the call to
+ * the fetcher function will suspend until the counter is reset. When
+ * this happens, an event is raised. In this system, we use this event
+ * to mark a sync run as "rate-limited", allowing us to complete the run
+ * and resume fetching the remaining data on the next run, when we will
+ * reset the counter.
+ */
+class LimitedFetcher extends EventEmitter {
+  // The maximum number of requests to allow
+  private limit: number;
+  // The number of requests made so far
+  private count: number;
+
+  public constructor(limit: number) {
+    super();
+    this.limit = limit;
+    this.count = 0;
+  }
+
+  public reset(): void {
+    this.count = 0;
+    this.emit("reset");
+  }
 
+  // Returns a function which matches the signature of the 'fetch' API
+  public fetcher(): FetchFn {
+    return async (input, init?) => {
+      if (this.count >= this.limit) {
+        // We are stuck at the limit, so do nothing until reset
+        // By awaiting here, we will pause indefinitely, relying on
+        // another system to reset our counter. In the sync context, the
+        // counter will be reset on each run, allowing us to pick up the
+        // request we were limited from making on the previous run.
+        this.emit("pause");
+        await new Promise<void>((resolve) => {
+          this.once("reset", () => {
+            resolve();
+          });
+        });
+      }
+      this.count++;
+      return fetch(input, init);
+    };
+  }
+
+  public getCount(): number {
+    return this.count;
+  }
+}
+
+/**
+ * For the purposes of sync, organizers are isolated from each other.
+ * A failure for one organizer should not prevent other organizers from
+ * completing sync. Organizers also have independent rate limits with
+ * Pretix.
+ */
+export class OrganizerSync {
```

Comment:
does an `OrganizerSync` instance live for longer than a single sync?

Reply:
Yes. Perhaps it makes sense to remove it once it's completed - it would be recreated automatically when the Pretix config is reloaded, which happens before each run. Would remove some of the need for internal variables to be reset, which is what happens currently.